### PR TITLE
Support ONNX "ReduceMean" Layer

### PR DESCRIPTION
**Pull request by l-bat**
Related: #14812 
**Merge with extra:** opencv/opencv_extra#634

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2019r1:16.04
build_image:Custom Win=openvino-2019r1
build_image:Custom Mac=openvino-2019r1

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-2
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -397,11 +397,30 @@ void ONNXImporter::populateNet(Net dstNet)
...
         }
-        else if (layer_type == "GlobalAveragePool" || layer_type == "GlobalMaxPool")
+        else if (layer_type == "GlobalAveragePool" || layer_type == "GlobalMaxPool" || layer_type == "ReduceMean")
         {
+            CV_Assert(node_proto.input_size() == 1);
             layerParams.type = "Pooling";
-            layerParams.set("pool", layer_type == "GlobalAveragePool" ? "AVE" : "MAX");
-            layerParams.set("global_pooling", true);
+            layerParams.set("pool", layer_type == "GlobalAveragePool" || layer_type == "ReduceMean" ? "AVE" : "MAX");
+            layerParams.set("global_pooling", layer_type != "ReduceMean");
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp]: I think we don't need `global_pooling` if we have `kernel_size` below

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -397,11 +397,30 @@ void ONNXImporter::populateNet(Net dstNet)
...
+
+                MatShape inpShape = outShapes[node_proto.input(0)];
+                if (inpShape.size() != 4 && inpShape.size() != 5)
+                    CV_Error(Error::StsNotImplemented, "Unsupported input shape of reduce_mean operation.");
+
+                DictValue axes = layerParams.get("axes");
+                CV_Assert(axes.size() <= inpShape.size() - 2);
+                std::vector<int> kernel_size(inpShape.size() - 2, 1);
+                for (int i = 0; i < axes.size(); i++)
+                    kernel_size[axes.get<int>(i) - 2] = inpShape[axes.get<int>(i)];
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp]: I think that we need to check the exact values of axes. We support only 2 or 3 for 4D and 2, 3, 4 for 5D

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 3]
```diff
@@ -752,11 +771,20 @@ void ONNXImporter::populateNet(Net dstNet)
...
+            CV_Assert(0 <= axis && axis <= inpShape.size());
+
+            std::vector<int> outShape;
+            for (int i = 0; i < axis; i++) {
+                outShape.push_back(inpShape[i]);
+            }
+            outShape.push_back(1);
+            for (int i = axis; i < inpShape.size(); i++) {
+                outShape.push_back(inpShape[i]);
+            }
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp]:
```cpp
std::vector<int> outShape = inpShape;
outShape.insert(outShape.begin() + axis, 1);
```

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Please get this value once and reuse:

```cpp
int axis = axes.get<int>(i);
```

File: modules/dnn/test/test_onnx_importer.cpp [hunk 4]
```diff
@@ -149,6 +149,14 @@ TEST_P(Test_ONNX_layers, Clip)
     testONNXModels("clip", npy);
 }
 
+TEST_P(Test_ONNX_layers, ReduceMean)
+{
+    if (target != DNN_TARGET_CPU)
+        throw SkipTestException("Only CPU is supported");
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp]: Can you please check if nonn-CPU cases still do not work?

File: modules/dnn/src/layers/layers_common.cpp [hunk 5]
```diff
@@ -172,6 +170,7 @@ void getPoolingKernelParams(const LayerParams &params, std::vector<size_t>& kern
     {
         util::getKernelSize(params, kernel);
     }
+    util::getStrideAndPadding(params, pads_begin, pads_end, strides, padMode, !globalPooling? kernel.size() : 2);
```
Comment (dkurt) [inline modules/dnn/src/layers/layers_common.cpp]: I think we can keep it unchanged, can you check?

Comment (l-bat) [reply modules/dnn/src/layers/layers_common.cpp]: No, we need to know kernel.size to set correct pads and strides size

File: modules/dnn/test/test_onnx_importer.cpp (hunk 4 above)
Comment (l-bat) [reply modules/dnn/test/test_onnx_importer.cpp]: reduce_mean3d (avepool3d) works only on CPU

Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp]: I see. Can we split test in two?

**Comment by dkurt** [comment]
May I ask you to check failed test with Myriad? https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/2379. We can skip it now or try to fix if it's possible.
```
[ RUN      ] Test_TensorFlow_layers.Split/1, where GetParam() = DLIE/MYRIAD
/build/precommit_custom_linux/3.4/opencv/modules/dnn/test/test_common.impl.hpp:63: Failure
Expected: (normL1) <= (l1), actual: 0.29697 vs 0.004
/build/precommit_custom_linux/3.4/opencv/modules/dnn/test/test_common.impl.hpp:66: Failure
Expected: (normInf) <= (lInf), actual: 1.20457 vs 0.02
```

**Pull request merged**